src/internal/clusterstate/v2.8.0/pfsdb.go

```diff
@@ -212,6 +212,103 @@ func migrateRepos(ctx context.Context, tx *pachsql.Tx) error {
+//   - branch_id_str is a metadata reference to the branches table. Today this points to collections, but tomorrow it should
+//     point to pfs.branches. Once the PFS master watches branches, we will no longer need this column.
+func alterCommitsTable(ctx context.Context, tx *pachsql.Tx) error {
+	query := `
+	CREATE TYPE pfs.commit_origin AS ENUM ('ORIGIN_KIND_UNKNOWN', 'USER', 'AUTO', 'FSCK');
+
+	ALTER TABLE IF EXISTS pfs.commits
+		ADD COLUMN IF NOT EXISTS repo_id bigint REFERENCES pfs.repos(id),
+		ADD COLUMN IF NOT EXISTS origin pfs.commit_origin,
+		ADD COLUMN IF NOT EXISTS description text DEFAULT '',
+		ADD COLUMN IF NOT EXISTS start_time timestamptz,
+		ADD COLUMN IF NOT EXISTS finishing_time timestamptz,
+		ADD COLUMN IF NOT EXISTS finished_time timestamptz,
+		ADD COLUMN IF NOT EXISTS compacting_time bigint,
+		ADD COLUMN IF NOT EXISTS validating_time bigint,
+		ADD COLUMN IF NOT EXISTS error text,
+		ADD COLUMN IF NOT EXISTS size bigint,
+		ADD COLUMN IF NOT EXISTS updated_at timestamptz DEFAULT CURRENT_TIMESTAMP,
+		ADD COLUMN IF NOT EXISTS branch_id_str text;
+
+	CREATE TRIGGER set_updated_at
+		BEFORE UPDATE ON pfs.commits
+		FOR EACH ROW EXECUTE PROCEDURE core.set_updated_at_to_now();
+	`
+	if _, err := tx.ExecContext(ctx, query); err != nil {
+		return errors.Wrap(err, "altering commits table")
+	}
+	return nil
+}
+
+// commitAncestry is how we model the commit graph within a single repo.
+func createCommitAncestryTable(ctx context.Context, tx *pachsql.Tx) error {
+	query := `
+	CREATE TABLE IF NOT EXISTS pfs.commit_ancestry (
+		from_id bigint REFERENCES pfs.commits(int_id),
+		to_id bigint REFERENCES pfs.commits(int_id),
+		PRIMARY KEY (from_id, to_id)
+	);
+	CREATE INDEX ON pfs.commit_ancestry (to_id, from_id);
+	CREATE INDEX ON pfs.commit_ancestry (to_id);
+	`
+	if _, err := tx.ExecContext(ctx, query); err != nil {
+		return errors.Wrap(err, "creating commit_ancestry table")
+	}
+	return nil
+}
+
+//nolint:unused //will use after table migration logic is implemented.
+func createNotifyCommitsTrigger(ctx context.Context, tx *pachsql.Tx) error {
+	query := `
+	CREATE OR REPLACE FUNCTION pfs.notify_commits() RETURNS TRIGGER AS $$
+	DECLARE
+		row record;
+		payload text;
+	BEGIN
+		IF TG_OP = 'DELETE' THEN
+			row := OLD;
+		ELSE
+			row := NEW;
+		END IF;
+		payload := TG_OP || ' ' || row.int_id::text;
+		PERFORM pg_notify('pfs_commits', payload);
+		PERFORM pg_notify('pfs_commits_repo_' || row.repo_id::text, payload);
+		return row;
+	END;
+	$$ LANGUAGE plpgsql;
+	CREATE TRIGGER notify
+		AFTER INSERT OR UPDATE OR DELETE ON pfs.commits
+		FOR EACH ROW EXECUTE PROCEDURE pfs.notify_commits();
+	`
+	if _, err := tx.ExecContext(ctx, query); err != nil {
+		return errors.Wrap(err, "creating notify trigger on pfs.commits")
+	}
+	return nil
+
+}
+
+// Migrate commits from collections.commits to pfs.commits
+func migrateCommits(ctx context.Context, tx *pachsql.Tx) error {
```

Comment:
I kinda like the way Bob implemented `synthesizeSpecs`, keeps the `clusterstate.go` cleaner.
```suggestion
func migrateCommits(ctx context.Context, env migrations.Env) error {
```